src/vcpkg/binarycaching.cpp

```diff
@@ -1255,6 +1255,159 @@ namespace
+                                     msg::url = docs::troubleshoot_binary_cache_url);
+                    }
+                    return LocalizedString::from_raw(std::move(res).output);
+                });
+        }
+        Path az_cli;
+    };
+
+    struct AzureUpkgPutBinaryProvider : public IWriteBinaryProvider
+    {
+        AzureUpkgPutBinaryProvider(const ToolCache& cache, MessageSink& sink, std::vector<AzureUpkgSource>&& sources)
+            : m_azure_tool(cache, sink), m_sources(sources)
+        {
+        }
+
+        size_t push_success(const BinaryPackageWriteInfo& request, MessageSink& msg_sink) override
+        {
+            size_t count_stored = 0;
+
+            auto package_version = format_version_for_nugetref(request.version.text, request.package_abi);
+            std::string package_name = fmt::format("{}-{}", request.spec.name(), request.spec.triplet());
+            std::string package_description = "Cached package for " + package_name;
+
+            for (auto&& write_src : m_sources)
+            {
+                auto res = m_azure_tool.publish(
+                    write_src, package_name, package_version, request.package_dir, package_description, msg_sink);
+                if (res)
+                {
+                    count_stored++;
+                }
+                else
+                {
+                    msg::println(res.error());
+                }
+            }
+
+            return count_stored;
+        }
+
+        bool needs_nuspec_data() const override { return false; }
+        bool needs_zip_file() const override { return false; }
+
+    private:
+        AzureUpkgTool m_azure_tool;
+        std::vector<AzureUpkgSource> m_sources;
+    };
+
+    struct AzureUpkgGetBinaryProvider : public IReadBinaryProvider
+    {
+        AzureUpkgGetBinaryProvider(const ToolCache& cache, MessageSink& sink, AzureUpkgSource source)
+            : m_azure_tool(cache, sink), m_sink(sink), m_source(source)
+        {
+        }
+
+        // Prechecking doesn't exist with universal packages so it's not implemented
+        void precheck(View<const InstallPlanAction*>, Span<CacheAvailability>) const override { }
+
+        LocalizedString restored_message(size_t count,
+                                         std::chrono::high_resolution_clock::duration elapsed) const override
+        {
+            return msg::format(msgRestoredPackagesFromAZUPKG, msg::count = count, msg::elapsed = ElapsedTime(elapsed));
+        }
+
+        void fetch(View<const InstallPlanAction*> actions, Span<RestoreResult> out_status) const override
+        {
+            for (size_t i = 0; i < actions.size(); ++i)
+            {
+                auto info = BinaryPackageReadInfo{*actions[i]};
+                auto package_version = format_version_for_nugetref(info.version.text, info.package_abi);
+                std::string package_name = fmt::format("{}-{}", info.spec.name(), info.spec.triplet());
```

Comment:
If this can't be `spec.to_string()` for some reason it should still be extracted as a separate function as `AzureUpkgPutBinaryProvider` and `AzureUpkgGetBinaryProvider` need to agree.

Reply:
I've rewritten this section to reuse what is used for Nuget binary caching, and renamed some functions.